### ISSUE TITLE
Dragging a selection does not adhere to `snapGrid`

### ISSUE THREAD

**Issue by louis-cf-lin** [wbkd/react-flow#1395]
Making an area selection and dragging it around doesn't cause it to snap to the specified grid size, hence on release, the selection can be dropped in misaligned positions.

**Comment by rickarubio** [comment]
I'm also seeing this issue, particularly if I shift + click to draw a selection around 2 nodes so that they are both selected.

Then I zoom in as far as I can.

At this point, I can then drag my selection in smaller increments than my grid settings, which causes items moved in this way to be misaligned.

The issue only seems to happen on multi-select + zooming in -- I've been poking around the useZoomPanHelpers but if anyone knows of a better starting point for investigating this issue please let me know. Thanks!

**Comment by rickarubio** [comment]
Actually, yeah I see the same issue if I shift + click and select even just 1 node and zoom all the way in afterwards, then attempt to drag the node along the x-axis.

I'm able to move the node around in smaller increments than my snapGrid settings.

**Comment by louis-cf-lin** [comment]
In my case, it happens even when my `zoom=1` (default). I haven't looked into the source code as I don't have the skills to do that sort of debugging but I've got a temporary workaround:

```javascript
const getNearestGridPosition = (position) => {
  const gridSize = 16
  return Math.round(position / gridSize) * gridSize ;
};

const selectionDragStopHandler = (_, selectionNodes) => {
  props.setElements((els) => {
    return els.map((el) => {
      for (const node of selectionNodes) {
        if (node.id === el.id) {
          return {
            id: el.id,
            type: el.type,
            position: {
              x: getNearestGridPosition(node.position.x),
              y: getNearestGridPosition(node.position.y),
            },
            data: el.data,
          };
        }
      }
      return el;
    });
  });
};
...
<ReactFlow onSelectionDragStop={selectionDragStopHandler}></ReactFlow>
```

Of course, this comes with the caveat that positions are stored in the `elements` state but I have that for other features anyways. Also, this is probably not the most efficient implementation since it iterates through all the selection nodes for every existing element in the `elements` state. Hope this helps!

**Comment by rickarubio** [comment]
Upon further investigation in `src/components/NodesSelection/index.tsx` it seems that whatever snapGrid is set to, when you drag a selection, it will move in half-increments.

For example, set snapGrid to [200, 200], shift + click to select 1 node, panning/zoom in / zoom out doesn't matter, just try drag it along the x-axis. If you move one increment to the right, it will be a 100 point shift on the x-axis.

Try with [400, 400] snapGrid, and the first increment will be 200.

For now, I have a little workaround:

Inside of ` src/components/NodesSelection/index.tsx`:
```
const onDrag = useCallback(
    (event: MouseEvent, data: DraggableData) => {
    console.log('dragging!!!!!!!');
      if (onSelectionDrag) {
        onSelectionDrag(event, selectedNodes);
      }

      console.log('X: ', data.deltaX);
      console.log('Y: ', data.deltaY);
      updateNodePosDiff({
        diff: {
          x: 2 * data.deltaX,
          y: 2 * data.deltaY,
        },
        isDragging: true,
      });
    },
    [onSelectionDrag, selectedNodes, updateNodePosDiff]
  );
```

In the call to `updateNodePosDiff` I am add a 2x multiplier to data.deltaX and data.deltaY.

However the selection rectangle is not properly offset as I'm still trying to figure out what controls its location.

![snapGridPartialFix](https://user-images.githubusercontent.com/3452670/127723849-52674753-9e84-4e65-a52e-99ecf2e4df1d.gif)

**Comment by louis-cf-lin** [comment]
Thanks for looking into it and finding where the problems begin. I'll keep my workaround for now since it's only triggered at the end of a drag so there's no issues with the blue selection rectangle becoming desynced. And also since it works for selection drag which is the main cause of this problem. Let me know if you do find how the blue rectangle is rendered though!

**Comment by moklick** [comment]
fix released in v10.3.0
